homeassistant/components/ollama_conversation/__init__.py

```diff
@@ -0,0 +1,263 @@
+    """Ollama conversation agent."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the agent."""
+        self.hass = hass
+        self.entry = entry
+
+        # conversation id -> message history
+        self._history: dict[str, MessageHistory] = {}
+
+    @property
+    def supported_languages(self) -> list[str] | Literal["*"]:
+        """Return a list of supported languages."""
+        return MATCH_ALL
+
+    async def async_process(
+        self, user_input: conversation.ConversationInput
+    ) -> conversation.ConversationResult:
+        """Process a sentence."""
+        settings = {**self.entry.data, **self.entry.options}
+
+        client = self.hass.data[DOMAIN][self.entry.entry_id]
+        conversation_id = user_input.conversation_id or ulid.ulid_now()
+        model = settings[CONF_MODEL]
+
+        # Look up message history
+        message_history: MessageHistory | None = None
+        message_history = self._history.get(conversation_id)
+        if message_history is None:
+            # New history
+            #
+            # Render prompt and error out early if there's a problem
+            raw_prompt = settings[CONF_PROMPT]
+            try:
+                prompt = self._generate_prompt(raw_prompt)
+                _LOGGER.debug("Prompt: %s", prompt)
+            except TemplateError as err:
+                _LOGGER.error("Error rendering prompt: %s", err)
+                intent_response = intent.IntentResponse(language=user_input.language)
+                intent_response.async_set_error(
+                    intent.IntentResponseErrorCode.UNKNOWN,
+                    f"Sorry, I had a problem generating my prompt: {err}",
+                )
+                return conversation.ConversationResult(
+                    response=intent_response, conversation_id=conversation_id
+                )
+
+            message_history = MessageHistory(
+                timestamp=time.monotonic(),
+                messages=[
+                    ollama.Message(role=MessageRole.SYSTEM.value, content=prompt)
+                ],
+            )
+            self._history[conversation_id] = message_history
+        else:
+            # Bump timestamp so this conversation won't get cleaned up
+            message_history.timestamp = time.monotonic()
+
+        # Clean up old histories
+        self._prune_old_histories()
+
+        # Trim this message history to keep a maximum number of *user* messages
+        max_messages = int(settings.get(CONF_MAX_HISTORY, MAX_HISTORY_NO_LIMIT))
+        self._trim_history(message_history, max_messages)
+
+        # Add new user message
+        message_history.messages.append(
+            ollama.Message(role=MessageRole.USER.value, content=user_input.text)
+        )
+
+        # Get response
+        try:
+            response = await client.chat(
+                model=model,
+                messages=list(message_history.messages),
```

Comment:
Why copy this ?

Reply:
With this, in the tests the mocked `client.chat` call args all had the same value because I append to the message history after the call. So every `call_args["messages"]` had the full history instead of just the history up to that point.

I figured this would be better than working around it, since the Ollama client could technically store a reference to the list for some reason.